tfsdk/block_test.go

```diff
@@ -5,13 +5,765 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	testtypes "github.com/hashicorp/terraform-plugin-framework/internal/testing/types"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
 	"github.com/hashicorp/terraform-plugin-go/tftypes"
 )
 
+func TestBlockModifyPlan(t *testing.T) {
+	t.Parallel()
+
+	schema := func(blockPlanModifiers AttributePlanModifiers, nestedAttrPlanModifiers AttributePlanModifiers) Schema {
+		return Schema{
+			Blocks: map[string]Block{
+				"test": {
+					Attributes: map[string]Attribute{
+						"nested_attr": {
+							Type:          types.StringType,
+							Required:      true,
+							PlanModifiers: nestedAttrPlanModifiers,
+						},
+					},
+					NestingMode:   BlockNestingModeList,
+					PlanModifiers: blockPlanModifiers,
+				},
+			},
+		}
+	}
+
+	schemaTfValue := func(nestedAttrValue string) tftypes.Value {
+		return tftypes.NewValue(
+			tftypes.Object{
+				AttributeTypes: map[string]tftypes.Type{
+					"test": tftypes.List{
+						ElementType: tftypes.Object{
+							AttributeTypes: map[string]tftypes.Type{
+								"nested_attr": tftypes.String,
+							},
+						},
+					},
+				},
+			},
+			map[string]tftypes.Value{
+				"test": tftypes.NewValue(
+					tftypes.List{
+						ElementType: tftypes.Object{
+							AttributeTypes: map[string]tftypes.Type{
+								"nested_attr": tftypes.String,
+							},
+						},
+					},
+					[]tftypes.Value{
+						tftypes.NewValue(
+							tftypes.Object{
+								AttributeTypes: map[string]tftypes.Type{
+									"nested_attr": tftypes.String,
+								},
+							},
+							map[string]tftypes.Value{
+								"nested_attr": tftypes.NewValue(tftypes.String, nestedAttrValue),
+							},
+						),
+					},
+				),
+			},
+		)
+	}
+
+	var schemaNullTfValue tftypes.Value = tftypes.NewValue(
```

Comment:
Tiny nitpick:

if `schemaTfValue` accepted a `tftypes.Value` that it used as the value of `nested_attr`, we could avoid having two of these helpers while still avoiding repeating large, large blocks of verbose object construction.

Not a big deal, if you don't like it as much we don't need to do it, just mentioning it.

Reply:
Sorry just to clarify, in `schemaTfValue` are you thinking that we should pass in the list value for each invocation? e.g. this part

```go
					[]tftypes.Value{
						tftypes.NewValue(
							tftypes.Object{
								AttributeTypes: map[string]tftypes.Type{
									"nested_attr": tftypes.String,
								},
							},
							map[string]tftypes.Value{
								"nested_attr": tftypes.NewValue(tftypes.String, nestedAttrValue),
							},
						),
					},
```

It's already accepting the string value for `nested_attr`. The `schemaNullTfValue` helper is handling where that list value is `nil`. Just trying to make sure I understand, thanks!

Reply:
You understand, I'm just confused.

I apparently briefly thought the `tftypes.String` in `nested_attr` was getting set to `null`, which is not what's happening here at all. I'm just being wrong.

Sorry, ignore this.